pkg/build/controller/buildimagechangetrigger/buildimagechangetrigger_controller.go

```diff
@@ -0,0 +1,293 @@
+
+	c := &BuildImageChangeTriggerController{
+		bcInstantiator: buildclient.NewOSClientBuildConfigInstantiatorClient(oc),
+		queue:          workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	c.bcStore.Indexer = bcInformer.GetIndexer()
+	c.isStore.Indexer = isInformer.GetIndexer()
+	isInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addImageStream,
+		UpdateFunc: c.updateImageStream,
+	})
+
+	c.bcStoreSynced = bcInformer.HasSynced
+	c.isStoreSynced = isInformer.HasSynced
+
+	return c
+}
+
+func (bc *BuildImageChangeTriggerController) addImageStream(obj interface{}) {
+	stream := obj.(*imageapi.ImageStream)
+	glog.V(4).Infof("Image stream %q added.", stream.Name)
+	bc.enqueueImageStream(stream)
+}
+
+func (bc *BuildImageChangeTriggerController) updateImageStream(old, cur interface{}) {
+	// A periodic relist will send update events for all known streams.
+	newStream := cur.(*imageapi.ImageStream)
+	oldStream := old.(*imageapi.ImageStream)
+	if newStream.ResourceVersion == oldStream.ResourceVersion {
+		return
+	}
+
+	glog.V(4).Infof("Image stream %q updated.", newStream.Name)
+	bc.enqueueImageStream(newStream)
+}
+
+func (bc *BuildImageChangeTriggerController) enqueueImageStream(stream *imageapi.ImageStream) {
+	key, err := kcontroller.KeyFunc(stream)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", stream, err))
+		return
+	}
+	bc.queue.Add(key)
+}
+
+func (c *BuildImageChangeTriggerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Wait for the bc store to sync before starting any work
+	if !cache.WaitForCacheSync(stopCh, c.bcStoreSynced, c.isStoreSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down build image change trigger controller")
+}
+
+func (bc *BuildImageChangeTriggerController) worker() {
+	for {
+		if quit := bc.work(); quit {
+			return
+		}
+	}
+}
+
+func (bc *BuildImageChangeTriggerController) work() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return true
+	}
+	defer bc.queue.Done(key)
+
+	is, err := bc.getImageStreamByKey(key.(string))
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+
+	if is == nil {
+		return false
+	}
+
+	err = bc.HandleImageStream(is)
+	bc.handleError(err, key, is)
+
+	return false
+
+}
+
+func (bc *BuildImageChangeTriggerController) getImageStreamByKey(key string) (*imageapi.ImageStream, error) {
+	obj, exists, err := bc.isStore.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve imagestream %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return nil, err
+	}
+	if !exists {
+		glog.Infof("ImageStream %q has been deleted", key)
```

Comment:
deleted is an overstatement here.  all we know is it's not in the cache for some reason.  Admittedly for the flow you're doing here, i think it's impossible for it not to be in the cache unless it's been deleted.  (since this flow was kicked off by an informer event for the imagestream, which means the imagestream was in the cache at that point).